### ISSUE TITLE
Missing tool: wacky brush, 3d brush

### ISSUE THREAD

**Issue by gingerbeardman** [vikrum/kidpix#9]
## Your sample image
![https://user-images.githubusercontent.com/291215/128236489-32a56b21-17e9-4b4a-b631-a987472127c4.png](https://user-images.githubusercontent.com/291215/128236489-32a56b21-17e9-4b4a-b631-a987472127c4.png)

It seems to me that this is drawing two strokes at once.

1. solid line
2. dither highlight

## More Examples

**Here are two ovals:**

1. left: drawn clockwise
2. right: drawn anti-clockwise

These were drawn starting/ending at the 12 o'clock position,
but the result is the same regardless of starting/ending position.

<img width="383" alt="Screen shot 2021-08-07 at 14 21 38" src="https://user-images.githubusercontent.com/49612/128601655-ac7ce630-087c-4ef2-9cf2-7b84f5fc4d09.png">

Notes
- they are diagonally mirrored along the line top-left to bottom-right

----

**Here are some horizontal and vertical lines:**

1. drawn upwards (bottom to top)
2. draw downwards (top to bottom)
3. drawn rightwards (left to right)
4. drawn leftwards (right to left)

<img width="200" alt="Screen shot 2021-08-07 at 14 24 54" src="https://user-images.githubusercontent.com/49612/128601726-9bd46686-4c23-49ed-a523-182d73b7cf35.png">

----

**Here are some diagonal lines:**

1. drawn upwards to left (bottom to top)
2. draw downwards to right (top to bottom)
3. drawn upwards to right (bottom to top)
4. draw downwards to left (top to bottom)

<img width="286" alt="Screen shot 2021-08-07 at 14 29 57" src="https://user-images.githubusercontent.com/49612/128601862-ecfd1dfc-e46a-495e-8719-8c940860f978.png">

Notes
- 1 & 2 are "opposite" (let's say of widths 0 and 1)
- 3 & 4 are "identical" (let's say both of width 2)

----

**A reproduction of your ornate shape:**

- drawn starting from A and ending at B

<img width="297" alt="Screen shot 2021-08-07 at 14 36 47" src="https://user-images.githubusercontent.com/49612/128602055-8b652ef9-1f83-478f-bb70-bdf4b07dd88b.png">

----

## My educated guess

Here's how I _think_ it does it:

1. draw the solid stroke
2. set size of the dither stoke according to direction (delta from last draw)
3. draw the dither stroke up-left a few pixels

### Size of dither stroke
- up/left decreases the size
- down/right increases the size

But from what default values?
And by what factor?

I'd have to experiment with code to figure out the details.

**Comment by vikrum** [comment]
Thank you for this! A kind soul (isachs@) reached out to Craig Hickman about the 3d brush tool and he responded back with this:

>  "I didn’t know about this. It’s an interesting implementation. The line he has questions about is simply two lines drawn at the same time. One solid on the bottom and a line with a 50% halftone pattern on the top. The first version of Kid Pix was written for the original Macintosh one bit graphics."

I'll keep noodling on it too. Worst comes to worst, the freeware/public domain version of KidPix is available here: http://cd.textfiles.com/garbo/MAC/GRAPHICS/ and we can reverse engineer that.

**Comment by gingerbeardman** [comment]
So I was right! That's good to know. 

That said, neither Craig or myself have the exact details just yet. So close!

I'll keep thinking about it.

**Comment by ubuntor** [comment]
reverse engineered the public domain version a bit, and i think it goes like this:

while mouse held down:
- set pen size to 5x5
- set pattern to solid
- draw line from `(old_x, old_y)` to `(new_x-1, new_y-1)`
- set pen size to 2x2
- set pattern to 50% black (checkerboard)
- draw line from `(old_x, old_y)` to `(new_x-2, new_y-2)`

**Comment by gingerbeardman** [comment]
Wow! Thanks @ubuntor for your RE work on this.

I'm amazed that got very close to how it works just by eyeballing.

**Comment by vikrum** [comment]
@ubuntor Huge thank you for digging this out. After giving it a try and not getting it to replicate, I figured out why both @gingerbeardman and I went down the wrong path (my guess too was that there was some implementation detail factoring in deltas or direction). 

Turns out I was confusing myself with the velocity of the brush! It's an implementation detail that in the OG version the _direction_ of the strokes causes previous renderings to stomp on each other! That's a huge part of the effect! If you increase the speed and try all the different directions, orientations, the dithered 2nd line shows up! 😂  

See towards end of video here:
https://user-images.githubusercontent.com/291215/129132915-67a5e714-137b-455f-8fa9-78e4ff086246.mov

Rough ballpark in codepen:
![Screen Shot 2021-08-11 at 8 17 07 PM](https://user-images.githubusercontent.com/291215/129133150-98c959fe-42d0-4940-b627-63e47d174fc9.png)

Thanks @gingerbeardman @ubuntor ! I'll try to work on a release this weekend and get a bunch of other fixes out too.
